---
title: "R Notebook"
author: ansintsova
date: 09.12.21
output: html_notebook
---


```{r}
#install.packages("devtools")
#devtools::install_github("thomasp85/patchwork")
```


```{r}
#library(knitr)
library(tidyverse)
library(ggplot2)
library(patchwork)
```

Load annotated results.

```{r}
#setwd("code/notebooks/10_21_GO_heatmaps")
results <- read.csv("../../../data/12_21_results/06-11-2021-final-results-annotated.csv", row.names=1)
results %>% head()
```


1. Set up a function to get specific pathway/module
2. Set up a function to make a graph from that df


```{r}
draw_path<- function(df, term, term_col, title){
   path.df <- df %>% filter(grepl(term, !! sym(term_col)))
   path.df <- path.df %>% arrange(desc(start)) %>% mutate(gene=factor(gene, levels=unique(gene))) %>% group_by(gene, day) %>% summarise(medCI = log2(median(ci_nov, na.rm=TRUE)), hits=ifelse(sum(ifelse(padj_nov<0.05, 1,0)) > 0, 'black','white'), medZscore = median(zscore_nov, na.rm=TRUE))
  minval <- floor(path.df$medZscore %>% min())
  maxval <- ceiling(path.df$medZscore %>% max())
  g <- path.df %>% ggplot(aes(day, gene)) +
  geom_point(aes (color=medZscore), shape=15, size=5)+
    geom_point(shape = 0,alpha=1, size = 6,colour = path.df$hits)+
  labs(x = NULL, y = NULL, col = "Median Z-score", title=title, fontsize=8)+
  theme_classic() +
  scale_color_gradient2(mid="#FBFEF9",low="#0C6291",high="#A63446", limits=c(minval, maxval))+
    coord_fixed(ratio=0.8)
  g
}
```



```{r}
glyc <- draw_path(results, 'ko00010', 'KEGG_Pathway', 'Glycolysis')
tca <- draw_path(results, 'ko00020', 'KEGG_Pathway', 'TCA')
anresp <- draw_path(results, 'GO:0009061', 'GOs', 'Anaerobic Respiration')
aeresp <- draw_path(results, 'GO:0009060', 'GOs', 'Aerobic Respiration')
ppp <- draw_path(results, 'GO:0006098', 'GOs', 'Pentose Phosphate')
```



```{r}
tcap = tca/ppp
tcap
```
```{r}
metabolism <- glyc|(anresp/aeresp)|tcap
```

```{r}
ggsave("../../../data/12_21_results/metabolism.png", metabolism, dpi=300, width = 12, height=12)
```


```{r}
bioAA <- draw_path(results, 'ko01230', 'KEGG_Pathway', 'Biosynthesis of Amino Acids')
pur <- draw_path(results, 'ko00230', 'KEGG_Pathway', 'Purine Metabolism')
pyr <- draw_path(results, 'ko00240', 'KEGG_Pathway', 'Pyrimidine Metabolism')
faa <- draw_path(results, 'ko01212', 'KEGG_Pathway', 'Fatty Acids Metabolism')


```
- Plot stat sig kegg pathways
```{r}
lps <- draw_path(results, 'ko00540', 'KEGG_Pathway', 'Lipopolysaccharide biosynthesis')
ko01130 <-draw_path(results, 'ko01130', 'KEGG_Pathway', 'Biosynthesis of Antibiotics')
#ko00230
ko00520 <- draw_path(results, 'ko00520', 'KEGG_Pathway', 'Amino sugar and nucleotide sugar metabolism')
ko01110 <- draw_path(results, 'ko01110', 'KEGG_Pathway', 'Biosynthesis of secondary metabolites')
ko00521 <- draw_path(results, 'ko00521', 'KEGG_Pathway', 'Streptomycin biosynthesis')

#ko00521
#ko00020 # day 4
#ko00010 # day 3
#

```
```{r}
lps_related <- (lps/ko00521)|ko00520

```
```{r}
ggsave("../../../data/12_21_results/lps.png", lps_related, dpi=300, width = 12, height=12)
```
```{r}
path.df <- results %>% filter(grepl('ko01230', !! sym('KEGG_Pathway')))
path.df <- path.df %>% arrange(desc(start)) %>% mutate(gene=factor(gene, levels=unique(gene))) %>% group_by(gene, day) %>% summarise(medCI = log2(median(ci_nov, na.rm=TRUE)), hits=ifelse(sum(ifelse(padj_nov<0.05, 1,0)) > 0, 'black','white'), medZscore = median(zscore_nov, na.rm=TRUE))



baa1.df <- path.df[c(1:151),]
baa2.df <- path.df[c(152:nrow(path.df)), ]

minval <- floor(baa1.df$medZscore %>% min())
maxval <- ceiling(baa1.df$medZscore %>% max())

baa1 <- baa1.df %>% ggplot(aes(day, gene)) +
     geom_point(aes (color=medZscore), shape=15, size=5)+
     geom_point(shape = 0,alpha=1, size = 6,colour = baa1.df$hits)+
     labs(x = NULL, y = NULL, col = "Median Z-score", title="Amino Acid Biosynthesis", fontsize=8)+
     theme_classic() +
     scale_color_gradient2(mid="#FBFEF9",low="#0C6291",high="#A63446", limits=c(minval, maxval))+
     coord_fixed(ratio=0.8)

minval <- floor(baa2.df$medZscore %>% min())
maxval <- ceiling(baa2.df$medZscore %>% max())

baa2 <- baa2.df %>% ggplot(aes(day, gene)) +
     geom_point(aes (color=medZscore), shape=15, size=5)+
     geom_point(shape = 0,alpha=1, size = 6,colour = baa2.df$hits)+
     labs(x = NULL, y = NULL, col = "Median Z-score", title="Amino Acid Biosynthesis", fontsize=8)+
     theme_classic() +
     scale_color_gradient2(mid="#FBFEF9",low="#0C6291",high="#A63446", limits=c(minval, maxval))+
     coord_fixed(ratio=0.8)

```



```{r}
vits <- c('thiM', 'thiC', 'bioB', 'bioF', 'bioDb', 'ybiV(2)', 'ybiV(1)', 'pdxAa', 'pdxB')
path.df <- results %>% filter(gene %in% vits)

path.df <- path.df %>% arrange(desc(start)) %>% mutate(gene=factor(gene, levels=unique(gene))) %>% group_by(gene, day) %>% summarise(medCI = log2(median(ci_nov, na.rm=TRUE)), hits=ifelse(sum(ifelse(padj_nov<0.05, 1,0)) > 0, 'black','white'), medZscore = median(zscore_nov, na.rm=TRUE))

minval <- floor(path.df$medZscore %>% min())
maxval <- ceiling(path.df$medZscore %>% max())
cofac <- path.df %>% ggplot(aes(day, gene)) +
     geom_point(aes (color=medZscore), shape=15, size=5)+
     geom_point(shape = 0,alpha=1, size = 6,colour = path.df$hits)+
     labs(x = NULL, y = NULL, col = "Median Z-score", title="Vitamin/Co-factor Biosynthesis", fontsize=8)+
     theme_classic() +
     scale_color_gradient2(mid="#FBFEF9",low="#0C6291",high="#A63446", limits=c(minval, maxval))+
     coord_fixed(ratio=0.8)

```

```{r}
metabolism2 <- baa1|baa2|pur|(pyr/faa/cofac)

ggsave("../../../data/12_21_results/metabolism2_z.png", metabolism2, dpi=150, height=15, width=15)
```
Click the **Knit and Open Document** to built and preview an output.